docs/ksflatten.1
.TH "KSFLATTEN" "1"
.SH "NAME"
ksflatten \(em flatten a kickstart file
.SH "SYNOPSIS"
.PP
\fBksflatten\fR [\fB\-c\fR | \fB\-\-config INFILE\fP]  [\fB\-o\fR | \fB\-\-output OUTFILE\fP]  [\fB\-v\fR | \fB\-\-version VERSION\fP]
.SH "DESCRIPTION"
.PP
\fBksflatten\fR is a program that takes an input kickstart file potentially containing includes and generates an
output file with all the includes inlined.
.SH "EXIT STATUS"
.PP
\fBksflatten\fR returns 0 on success, and 1 if INFILE does not exist or there is an error parsing the kickstart file.
.SH "OPTIONS"
.IP "\fB\-c\fR, \fB\-\-config INFILE\fP" 10
The name of the input kickstart file.
.IP "\fB\-o\fR, \fB\-\-output OUTFILE\fP" 10
Write the flattened kickstart file to OUTFILE, or stdout if no filename is given.
.IP "\fB\-v\fR, \fB\-\-version VERSION\fP" 10
Use this version of kickstart syntax when processing the file, or the latest if no version is given.
.SH "SEE ALSO"
.PP
ksshell (1), ksvalidator (1), ksverdiff (1)


ChangeLog
2008-04-16  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* Makefile:  Update for git.

	* pykickstart/parser.py (KickstartHandlers.doNetwork): query is now a valid
	bootproto for networking (#441837).

2006-12-15  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version (only translations have
	changed since the last rebuild - #216620).

2006-12-05  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

2006-12-04  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter.doRepo):  Fix traceback
	(#218274).

2006-12-01  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/parser.py (KickstartHandlers.deprecatedCommands):  Fix
	traceback (#218047, #218059).

2006-11-30  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

2006-11-21  Chris Lumens  <clumens@redhat.com>

	* po/pykickstart.pot:  Updated (#216620).

	* pykickstart/parser.py:  Update string formats to make translation
	easier.

2006-11-14  Chris Lumens  <clumens@redhat.com>

	* pykickstart/parser.py (KickstartHandlers.doIscsi):  Use required
	attribute.

2006-11-09  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter.doLogging):  Add --level.

2006-10-24  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* Makefile:  Fix for new Release format.

2006-10-17  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py:  Add writer for key command.

2006-10-17  Jeremy Katz  <katzj@redhat.com>

	* pykickstart.spec, setup.py: Bump version.
	
	* pykickstart/parser.py, pykickstart/constants.py: add --skip for key

2006-10-16  Jeremy Katz  <katzj@redhat.com>

	* pykickstart.spec, setup.py: Bump version.

	* pykickstart/parser.py (KickstartHandlers.doKey): Support for
	installation key
	* pykickstart/data.py (KickstartData.__init__): Likewise.

2006-10-13  Bill Nottingham  <notting@redhat.com>

	* af.po, am.po, ar.po, be.po, bg.po, bn.po, cy.po, et.po,
	  eu_ES.po, fa.po, fi.po, he.po, hy.po, kn.po, ku.po, lo.po, lt.po,
	  lv.po, mk.po, my.po, nn.po, no.po, nso.po, ro.po, si.po, sq.po, th.po,
	  ur.po, vi.po, zu.po: Use a valid charset (#210720)
	
	* pykickstart.spec, setup.py: Bump version


2006-09-29  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py: Bump version

	* pykickstart/parser.py (KickstartHandlers.doMethod):  Fix traceback
	and test (#208557).

2006-09-25  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py: Bump version

	* pykickstart/parser.py (KickstartHandlers.doMethod):  Add support for
	--biospart option (#207585).
	* pykickstart/writer.py (KickstartWriter.doMethod):  Likewise.

	* pykickstart/writer.py (KickstartWriter.doIscsi):  Output new syntax.
	(KickstartWriter.doZFCP):  Likewise.
	(KickstartWriter.doIscsiName):  Added.

2006-09-20  Jeremy Katz  <katzj@redhat.com>

	* pykickstart.spec, setup.py: Bump version

	* pykickstart/parser.py (KickstartHandlers.doIscsi): New and
	improved iscsi syntax.  Also, allow multiple zfcp devs
	* pykickstart/data.py: Likewise.

2006-07-13  Chris Lumens  <clumens@redhat.com>

	* pykickstart/data.py:  Add noipv4 and noipv6 to KickstartNetworkData.
	* pykickstart/parser.py (KickstartHandlers.doNetwork):  Limit
	bootproto choices to what anaconda can handle.  Add support for
	--noipv4 and --noipv6 options.
	* pykickstart/writer.py (KickstartWriter.doNetwork):  Write out new
	networking options.

2006-06-20  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/writer.py (KickstartWriter.doIscsi):  Add iscsi writer.

2006-06-20  Jeremy Katz  <katzj@redhat.com>

	* pykickstart/parser.py (KickstartHandlers.doIscsi): Initial iscsi
	support.  This syntax is likely to change but this makes testing
	so much easier.
	* pykickstart/data.py (KickstartData.__init__): Add iscsi data.

2006-06-20  Chris Lumens  <clumens@redhat.com>

	* pykickstart/parser (KickstartHandlers.doRaid):  Handle a RAID device
	as an integer instead of a string (#176537).

2006-06-16  Jeremy Katz  <katzj@redhat.com>

	* pykickstart/parser.py (KickstartHandlers.doMethod): Handle nfs --opts

2006-06-06  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec:  Fix BuildRequires (#194156).

2006-05-25  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

2006-05-24  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter):  Change order of LVM-related
	functions (#193073).

2006-05-23  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec:  Require urlgrabber now.

	* po/*:  Update translation files.

	* pykickstart/parser.py (KickstartParser.handleCommand):  Return a
	more useful error message on unknown command.

	* pykickstart/writer.py (KickstartWriter.doLogicalVolume):  Fix typo.

	* validator/ksvalidator:  Accept a URL in addition to a file name.

2006-05-23  Chris Lumens  <clumens@redhat.com>

	* pykickstart/writer.py (KickstartWriter.doPackages):  Don't write out
	an empty packages section, since we may want this step to be
	interactive (#192851).

2006-05-23  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/writer.py (KickstartWriter.doFirewall):  Rename --ports
	to match the parser.

2006-05-17  Peter Jones  <pjones@redhat.com>

	* pykickstart/parser.py: Add "multipath" command and doMultiPath
	handler.
	* pykickstart/data.py: Add kickstartMpPathData and
	KickstartMultiPathData objects to handle multipath info, as well as
	KickstartData.mpaths .
	* pykickstart/writer.py: add doMultiPath
	* docs/programmers-guide: add completely useless documentation for
	multipath kickstart objects.

2006-05-15  Chris Lumens  <clumens@redhat.com>

	* pykickstart.spec, setup.py:  Bump version.

	* pykickstart/data.py (KickstartNetworkData):  Add MTU (#191328).

tools/ksflatten.py
#!/usr/bin/python3
#
# Simple script to take a kickstart config, read it in, parse any %includes,
# etc to write out a flattened config that is stand-alone
#
# Copyright 2007-2014, Red Hat, Inc.
# Jeremy Katz <katzj@redhat.com>
# Chris Lumens <clumens@redhat.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

from __future__ import print_function

import sys
import argparse
import pykickstart
import pykickstart.parser
from pykickstart.i18n import _
from pykickstart.version import DEVEL, makeVersion
from pykickstart.errors import KickstartVersionError

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("-c", "--config", dest="kscfg", required=True,
                        help=_("Path to kickstart config file"))
    parser.add_argument("-v", "--version", dest="version", default=DEVEL,
                        help=_("Kickstart version to use for interpreting config"))
    parser.add_argument("-o", "--output", dest="output",
                        help=_("Write flattened config to OUTPUT"))

    return parser.parse_args()

def main():
    opts = parse_args()
    if not opts.kscfg:
        print(_("Need to specify a config to flatten"), file=sys.stderr)
        sys.exit(1)

    try:
        ksversion = makeVersion(opts.version)
    except KickstartVersionError:
        print(_("The version %s is not supported by pykickstart") % opts.version)
        sys.exit(1)

    ksparser = pykickstart.parser.KickstartParser(ksversion)
    try:
        ksparser.readKickstart(opts.kscfg)
    except IOError as msg:
        print(_("Failed to read kickstart file '%(filename)s' : %(error_msg)s") % {"filename": opts.kscfg, "error_msg": msg}, file=sys.stderr)
        sys.exit(1)
    except pykickstart.errors.KickstartError as e:
        print(_("Failed to parse kickstart file '%(filename)s' : %(error_msg)s") % {"filename": opts.kscfg, "error_msg": e}, file=sys.stderr)
        sys.exit(1)

    if opts.output:
        try:
            f = open(opts.output, 'w')
        except IOError as msg:
            print(_("Failed to open output file '%(filename)s' : %(error_msg)s") % {"filename": opts.output, "error_msg": msg}, file=sys.stderr)
            sys.exit(1)
    else:
        f = sys.stdout

    f.write("%s" % ksparser.handler)
    f.close()

if __name__ == "__main__":
    main()
